onnxruntime/core/providers/cpu/math/top_heap.h

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include <vector>
+#include <thread>
+#include <iterator>
+#include <queue>
+#include <iostream>
+#include <algorithm>
+
+#if USE_OPENMP
+#include <omp.h>
+#endif
+
+namespace onnxruntime {
+
+int64_t flattened_dimension(const TensorShape& values) {
+  int64_t r = 1;
+  for (auto it = values.GetDims().begin(); it != values.GetDims().end(); ++it)
+    r *= *it;
+  return r;
+}
+
+void shape2strides(const TensorShape& shape, TensorShape& shape_strides) {
+  std::vector<int64_t> strides;
+  strides.resize(shape.NumDimensions());
+  strides[strides.size() - 1] = sizeof(int64_t);
+  for (int i = (int)strides.size() - 2; i >= 0; --i)
+    strides[i] = strides[i + 1] * shape[i + 1];
+  shape_strides = strides;
+}
+
+template <typename NTYPE>
+struct HeapMax {
+  using DataType = NTYPE;
+  bool cmp1(const NTYPE& v1, const NTYPE& v2) const {
+    return v1 > v2;
+  }
+  bool cmp(int64_t i1, int64_t i2, const NTYPE* ens, const int64_t* pos) const {
+    return (ens[pos[i1]] < ens[pos[i2]]) || ((pos[i1] > pos[i2]) && (ens[pos[i1]] == ens[pos[i2]]));
+  }
+};
+
+template <typename NTYPE>
+struct HeapMin {
+  using DataType = NTYPE;
+  bool cmp1(const NTYPE& v1, const NTYPE& v2) const {
+    return v1 < v2;
+  }
+  bool cmp(int64_t i1, int64_t i2, const NTYPE* ens, const int64_t* pos) const {
+    return (ens[pos[i1]] > ens[pos[i2]]) || ((pos[i1] > pos[i2]) && (ens[pos[i1]] == ens[pos[i2]]));
+  }
+};
+
+template <class HeapCmp>
+void _heapify_up_position(const typename HeapCmp::DataType* ens, int64_t* pos, size_t i, size_t k, const HeapCmp& heap_cmp) {
+  size_t left, right;
+  int64_t ch;
+  while (true) {
+    left = 2 * i + 1;
+    right = left + 1;
+    if (right < k) {
+      if (heap_cmp.cmp(left, i, ens, pos) && !heap_cmp.cmp1(ens[pos[left]], ens[pos[right]])) {
```

Comment:
Doesn't the comparison of left and right need to check the index as well in case the values are equal, so cmp not cmp1 should be used?